nixos/modules/services/web-apps/discourse.nix

```diff
@@ -0,0 +1,471 @@
+    };
+
+   systemd.services.discourse-sidekiq = {
+      description = "Discourse job queue";
+
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" "discourse.service" ];
+      environment = dcEnv;
+      path = with pkgs; [
+        procps # ps
+      ];
+      serviceConfig = dcServiceConfig // {
+        ExecStart = "${discourse-bundler}/bin/discourse-bundler exec sidekiq";
+      };
+    };
+
+    services.postgresql = optionalAttrs cfg.database.createLocally {
+      enable = true;
+      ensureDatabases = [ cfg.database.name ];
+      ensureUsers = [
+        { name = cfg.database.user;
+          ensurePermissions = {
+            "DATABASE ${cfg.database.name}" = "ALL PRIVILEGES";
+          };
+        }
+      ];
+    };
+
+    # The postgresql module doesn't currently support concepts like
+    # objects owners and extensions; for now we tack on what's needed
+    # here.
+    systemd.services.postgresql.postStart = mkAfter (optionalString cfg.database.createLocally ''
+      set -eu
+
+      $PSQL '${cfg.database.name}' -tAc "CREATE EXTENSION IF NOT EXISTS hstore"
+      $PSQL '${cfg.database.name}' -tAc "CREATE EXTENSION IF NOT EXISTS pg_trgm"
+    '');
+
+    services.redis.enable = mkDefault true;
+
+    services.postfix = mkIf cfg.mail.enable {
+      enable = true;
+
+      relayDomains = cfg.mail.domains;
+      transport = concatMapStringsSep "\n"
+        (dom: "${dom} discourse:") cfg.mail.domains;
+      config = {
+        smtpd_recipient_restrictions = "check_policy_service unix:private/policy";
+      };
+      masterConfig = {
+        "discourse" = {
+          type = "unix";
+          privileged = true;
+          chroot = false;
+          command = "pipe";
+          args = [
+            "user=nobody:nogroup"
+            "argv=${cfg.mail.package}/bin/receive-mail"
+            "\${recipient}"
+          ];
+        };
+        "policy" = {
+          type = "unix";
+          privileged = true;
+          chroot = false;
+          command = "spawn";
+          args = [
+            "user=nobody:nogroup"
+            "argv=${cfg.mail.package}/bin/discourse-smtp-fast-rejection"
+          ];
+        };
+      };
+    };
+
+    systemd.services.postfix.after = optionals
+      cfg.mail.enable [ "discourse-mail-init.service" ];
+
+    systemd.services.discourse-mail-init = mkIf cfg.mail.enable {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "discourse.service" ];
+
+      script = ''
+        apiKey="$(<'${cfg.mail.apiKeyFile}')"
+        echo '${ builtins.toJSON mailEnv }' | ${pkgs.jq}/bin/jq ".DISCOURSE_API_KEY=\"$apiKey\"" \
+          > ${dataDir}/mail-receiver-environment.json
+
+        chown postfix ${dataDir}/mail-receiver-environment.json
+        chmod 400 ${dataDir}/mail-receiver-environment.json
+      '';
+      serviceConfig.Type = "oneshot";
+    };
+
+    systemd.tmpfiles.rules = [
+      "C  ${dataDir} -    ${user} - - ${cfg.package}/share/discourse"
```

Comment:
Will upgrades work when old content is not overwritten by `C`?

Reply:
I'm not sure about this approach at  all, might be better to do this in `PreStart` as well.